lib/compat/wordpress-6.1/class-wp-theme-json-6-1.php

```diff
@@ -878,6 +878,41 @@ function( $pseudo_selector ) use ( $selector ) {
 			$block_rules .= static::to_ruleset( $feature_selector, $individual_feature_declarations );
 		}
 
+		// Add border color if we have a separator with a bg color defined.
+		if ( '.wp-block-separator' === $selector ) {
+			$border_color_matches = array_values(
+				array_filter(
+					$declarations,
+					function( $declaration ) {
+						return 'border-color' === $declaration['name'];
+					}
+				)
+			);
+			$text_color_matches   = array_values(
+				array_filter(
+					$declarations,
+					function( $declaration ) {
+						return 'color' === $declaration['name'];
+					}
+				)
+			);
+			$background_matches   = array_values(
+				array_filter(
+					$declarations,
+					function( $declaration ) {
+						return 'background-color' === $declaration['name'];
+					}
+				)
+			);
+			if ( ! empty( $background_matches && isset( $background_matches[0]['value'] ) ) && empty( $border_color_matches ) && empty( $text_color_matches ) ) {
+				$declarations[] = array(
+					'name'  => 'border-color',
+					'value' => $background_matches[0]['value'],
+				);
+				$block_rules    = static::to_ruleset( $selector, $declarations );
```

Comment:
> Finally, would it be worth moving this logic to a separate function, since it adds a fair bit to the length of get_styles_for_block? Or is it too specific to a single usage for it to be worth the abstraction 🤔

I tried to move it to a filter but I could not call the protected `to_ruleset` function. I also think is a good idea to move it, that way we could remove it in the future if we update the separator block to use only border options. 

Also, I thought to try the idea of just adding 1px height if a background is defined, that could do the work and would be less invasive, but I guess it won't work for dotted separators 😢 

I will update it with all the nits!

Reply:
> Should color also be output along with border color if color isn't already present in theme.json? As far as I know, it's only the dots style that is affected.

I think it could be better to use only color. We don't need border color, as it seems that color is a better fit (works for both line and dotted separators)

Reply:
> instead of appending to $declarations[], we could create a separate $declarations_separator array for this additional border-color rule, and use that directly in the ruleset output, a bit like how the duotone works. E.g. $block_rules .= static::to_ruleset( $selector_duotone, $declarations_separator );

I think just updating the declarations before step 2 will be more performant.